src/mds/QuiesceDb.h

```diff
@@ -0,0 +1,338 @@
+// The age of the database is just large enough to have earliest events carry
+// a non-negative age stamp.
+// This is sufficient because we only care to honor the timeouts that are relative
+// to the other recorded database events.
+// This approach also relieves us from storing or transfering absolute time stamps:
+// every client can deduce the lower boundary of event's absolute time given the 
+// message roundrip timing - if they bother enough. Otherwise, they can just subtract
+// the received database age from now() and get their own absolute time reference.
+
+using QuiesceClock = ceph::coarse_real_clock;
+using QuiesceTimePoint = QuiesceClock::time_point;
+using QuiesceTimeInterval = QuiesceClock::duration;
+using QuiesceSetId = std::string;
+using QuiesceRoot = std::string; // TODO: file path?
+using QuiesceDbVersion = uint32_t;
+
+struct RecordedQuiesceState {
+  QuiesceState state;
+  QuiesceTimeInterval at_age;
+
+  operator QuiesceState() {
+    return state;
+  }
+
+  bool update(const QuiesceState state, const QuiesceTimeInterval at_age) {
+    if (state != this->state) {
+      this->state = state;
+      this->at_age = at_age;
+      return true;
+    }
+    return false;
+  }
+
+  RecordedQuiesceState(QuiesceState state, QuiesceTimeInterval at_age) : state(state), at_age(at_age) {}
+  RecordedQuiesceState() : RecordedQuiesceState (QS__INVALID, QuiesceTimeInterval::zero()) {}
+};
+
+struct QuiesceSet {
+  struct MemberInfo {
+    RecordedQuiesceState rstate;
+    bool excluded = false;
+    MemberInfo(QuiesceState state, QuiesceTimeInterval at_age)
+        : rstate(state, at_age)
+        , excluded(false)
+    {
+    }
+    MemberInfo(QuiesceTimeInterval at_age)
+        : MemberInfo(QS_QUIESCING, at_age)
+    {
+    }
+
+    QuiesceState get_effective_state(QuiesceState parent_state = QS__INVALID) const { return excluded ? QS__INVALID : std::max(rstate.state, parent_state); }
+    bool is_quiesced(QuiesceState parent_state = QS__INVALID) const { return get_effective_state(parent_state) == QS_QUIESCED; }
+  };
+  QuiesceDbVersion version = 0;
+  RecordedQuiesceState rstate;
+  QuiesceTimeInterval timeout = QuiesceTimeInterval::zero();
+  QuiesceTimeInterval expiration = QuiesceTimeInterval::zero();
+  using Members = std::unordered_map<QuiesceRoot, MemberInfo>;
+  Members members;
+
+  bool is_active() const {
+    return
+      rstate.state > QS__INVALID
+      && rstate.state < QS__TERMINAL;
+  }
+
+  bool is_quiescing() const { return rstate.state == QS_QUIESCING; }
+  bool is_quiesced() const { return rstate.state == QS_QUIESCED; }
+  bool is_releasing() const { return rstate.state == QS_RELEASING; }
+  bool is_released() const { return rstate.state == QS_RELEASED; }
+};
+
+struct QuiesceDbRequest {
+  enum RootsOp {
+    INCLUDE_OR_QUERY,
+    EXCLUDE_OR_RELEASE,
+    RESET_OR_CANCEL
+  } roots_op = INCLUDE_OR_QUERY;
+  std::optional<std::string> set_id;
```

Comment:
explain why this is optional

Reply:
And what if e.g. timeout is omitted?

Reply:
Looking at the other members of this class one can follow the logic behind these.
Specifically, an optional set id is required to comply with our suggested API:

* including roots without providing a set id will generate a new set with a unique id
* querying without a set id will return the full db
* cancelling without a set id will cancel all active sets

Timeout and expiration are optional in a request, but if they are never provided for a new set, they stay 0. This will be obvious to the client because a new set with 0 timeout will enter the TIMEDOUT state immediately.
A set with 0 expiration makes more sense: it can be used as a barrier for synchronization.

All of the behaviors are validated by `SetCreation`, `SetModification`, and `MultiRankQuiesce`

Reply:
Added the appropriate comments to the QuiesceDb.h